src/flexmock/api.py

```diff
@@ -392,77 +381,66 @@ def _handle_matched_expectation(
+                raise return_value.raises(
+                    *return_value.value["kargs"], **return_value.value["kwargs"]
+                )
+            raise return_value.raises  # pylint: disable=raising-bad-type
 
         def mock_method(runtime_self: Any, *kargs: Any, **kwargs: Any) -> Any:
             arguments = {"kargs": kargs, "kwargs": kwargs}
             expectation = FlexmockContainer.get_flexmock_expectation(self, name, arguments)
             if expectation:
                 return _handle_matched_expectation(expectation, runtime_self, *kargs, **kwargs)
             # inform the user which expectation(s) for the method were _not_ matched
-            expectations = [
-                expectation
+            error_msg = _format_args(name, arguments) + "\n".join(
+                "\nDid not match expectation %s" % _format_args(name, expectation._args)
                 for expectation in reversed(FlexmockContainer.flexmock_objects.get(self, []))
                 if expectation.name == name
-            ]
-            error_msg = _format_args(name, arguments)
-            if expectations:
-                for expectation in expectations:
-                    error_msg += "\nDid not match expectation %s" % _format_args(
-                        name, expectation._args
-                    )
+            )
             # make sure to clean up expectations to ensure none of them
             # interfere with the runner's error reporting mechanism
             # e.g. open()
-            for _, expectations in FlexmockContainer.flexmock_objects.items():
-                for expectation in expectations:
-                    _getattr(expectation, "reset")()
+            for expectation in itertools.chain.from_iterable(
+                FlexmockContainer.flexmock_objects.values()
+            ):
+                expectation.reset()
             raise MethodSignatureError(error_msg)
 
         return mock_method
 
 
 def flexmock_teardown() -> None:
     """Performs flexmock-specific teardown tasks."""
-    saved = {}
-    instances = []
-    classes = []
-    for mock_object, expectations in FlexmockContainer.flexmock_objects.items():
-        saved[mock_object] = expectations[:]
-        for expectation in expectations:
-            _getattr(expectation, "reset")()
-    for mock in saved:
-        obj = mock._object
-        if not isinstance(obj, Mock) and not inspect.isclass(obj):
-            instances.append(obj)
-        if inspect.isclass(obj):
-            classes.append(obj)
-    for obj in instances + classes:
+    saved_flexmock_objects = FlexmockContainer.flexmock_objects
+    all_expectations = list(itertools.chain.from_iterable(saved_flexmock_objects.values()))
+    for expectation in all_expectations[:]:
+        expectation.reset()
+
+    def _is_instance_or_class(obj: Any) -> bool:
+        return bool(
+            (not isinstance(obj, Mock) and not inspect.isclass(obj)) or inspect.isclass(obj)
```

Comment:
Is that not already a boolean ? :thinking:

Reply:
It is but I guess it was done this way so that pylint doesn't complain about not using the ternary operator.